relay_client/src/http.rs

```diff
@@ -0,0 +1,281 @@
+        ConnectionOptions,
+        MessageIdGenerator,
+    },
+    http::{HeaderMap, StatusCode},
+    relay_rpc::{
+        auth::ed25519_dalek::Keypair,
+        domain::{DecodedClientId, SubscriptionId, Topic},
+        jwt::{self, JwtError, VerifyableClaims},
+        rpc::{self, RequestPayload},
+    },
+    std::{sync::Arc, time::Duration},
+    url::Url,
+};
+
+pub type TransportError = reqwest::Error;
+pub type Response<T> = Result<<T as RequestPayload>::Response, Error>;
+pub type EmptyResponse = Result<(), Error>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum HttpClientError {
+    #[error("HTTP transport error: {0}")]
+    Transport(#[from] TransportError),
+
+    #[error("Invalid request: {0}")]
+    InvalidRequest(BoxError),
+
+    #[error("Invalid response")]
+    InvalidResponse,
+
+    #[error("Invalid HTTP status: {0}")]
+    InvalidHttpCode(StatusCode),
+
+    #[error("JWT error: {0}")]
+    Jwt(#[from] JwtError),
+
+    #[error("RPC error: code={} message={}", .0.code, .0.message)]
+    RpcError(rpc::ErrorData),
+}
+
+#[derive(Debug, Clone)]
+pub struct WatchRegisterRequest {
+    /// Service URL.
+    pub service_url: String,
+    /// Webhook URL.
+    pub webhook_url: String,
+    /// Watcher type. Either subscriber or publisher.
+    pub watch_type: rpc::WatchType,
+    /// Array of message tags to watch.
+    pub tags: Vec<u32>,
+    /// Array of statuses to watch.
+    pub statuses: Vec<rpc::WatchStatus>,
+    /// TTL for the registration.
+    pub ttl: Duration,
+}
+
+#[derive(Debug, Clone)]
+pub struct WatchUnregisterRequest {
+    /// Service URL.
+    pub service_url: String,
+    /// Webhook URL.
+    pub webhook_url: String,
+    /// Watcher type. Either subscriber or publisher.
+    pub watch_type: rpc::WatchType,
+}
+
+/// The Relay HTTP RPC client.
+#[derive(Debug, Clone)]
+pub struct Client {
+    client: reqwest::Client,
+    url: Url,
+    origin: String,
+    id_generator: MessageIdGenerator,
+}
+
+impl Client {
+    pub fn new(opts: &ConnectionOptions) -> Result<Self, Error> {
+        let mut headers = HeaderMap::new();
+        opts.update_request_headers(&mut headers)?;
+
+        let client = reqwest::Client::builder()
+            .default_headers(headers)
+            .build()
+            .map_err(HttpClientError::Transport)?;
+
+        let url = opts.as_url()?;
+        let origin = url.origin().unicode_serialization();
+        let id_generator = MessageIdGenerator::new();
+
+        Ok(Self {
+            client,
+            url,
+            origin,
+            id_generator,
+        })
+    }
+
+    /// Publishes a message over the network on given topic.
+    pub async fn publish(
+        &self,
+        topic: Topic,
+        message: impl Into<Arc<str>>,
+        tag: u32,
+        ttl: Duration,
+    ) -> EmptyResponse {
+        self.request(rpc::Publish {
+            topic,
+            message: message.into(),
+            ttl_secs: ttl.as_secs() as u32,
```

Comment:
Either `#[allow(clippy::cast_possible_truncation)]` or `u32::try_from` should be used?